core/templates/pages/exploration-editor-page/services/changes-in-human-readable-form.service.spec.ts

```diff
@@ -0,0 +1,632 @@
+import { OutcomeObjectFactory } from
+  'domain/exploration/OutcomeObjectFactory';
+
+describe('Changes In Human Readable Form Service', () => {
+  let cihrfs, oof;
+
+  beforeEach(() => {
+    cihrfs = TestBed.get(ChangesInHumanReadableFormService);
+    oof = TestBed.get(OutcomeObjectFactory);
+  });
+
+  it('should make rules list human readable', () => {
+    const answerGroupValue = {
+      rules: [{
+        type: 'Type1',
+        inputs: {
+          input1: 'input1',
+          input2: 'input2'
+        }
+      }]
+    };
+
+    cihrfs.makeRulesListHumanReadable(answerGroupValue)
+      .forEach((element: HTMLElement, i) => {
+        const rule = answerGroupValue.rules[i];
+        expect(element.outerHTML).toBe(
+          '<li>' +
+          '<p>Type: ' + rule.type + '</p>' +
+          '<p>Value: ' + Object.keys(rule.inputs).toString() + '</p>' +
+          '</li>'
+        );
+      });
+  });
+
+  it('should not make rules list human readable when there are no rules',
+    () => {
+      const answerGroupValue = {
+        rules: []
+      };
+      expect(cihrfs.makeRulesListHumanReadable(answerGroupValue)).toEqual([]);
+    });
+
+  it('should get state property value when state is an array', () => {
+    const state = ['value1', 'value2'];
+    expect(cihrfs.getStatePropertyValue(state)).toBe('value2');
+  });
+
+  it('should get state property value when state is an object', () => {
+    const state = {
+      value1: ''
+    };
+    expect(cihrfs.getStatePropertyValue(state)).toBe(state);
+  });
+
+  it('should get relative change to groups when adding a new item in an array',
+    () => {
+      const changeObject = {
+        new_value: ['value1'],
+        old_value: []
+      };
+      expect(cihrfs.getRelativeChangeToGroups(changeObject)).toBe('added');
+    });
+
+  it('should get relative change to groups when editing an array', () => {
+    const changeObject = {
+      new_value: ['value2'],
+      old_value: ['value1']
+    };
+    expect(cihrfs.getRelativeChangeToGroups(changeObject)).toBe('edited');
+  });
+
+  it('should get relative change to groups when deleting items in an array',
+    () => {
+      const changeObject = {
+        new_value: [],
+        old_value: ['value1']
+      };
+      expect(cihrfs.getRelativeChangeToGroups(changeObject)).toBe('deleted');
+    });
+
+  it('should get relative change to groups when deleting properties in' +
+    ' an object', () => {
+    const changeObject = {
+      new_value: {},
+      old_value: {
+        property1: true
+      }
+    };
+    expect(cihrfs.getRelativeChangeToGroups(changeObject)).toBe('deleted');
+  });
+
+  it('should get relative change to groups when editing properties in' +
+    ' an object', () => {
+    const changeObject = {
+      new_value: {
+        property1: true
+      },
+      old_value: {
+        property1: true
+      }
+    };
+    expect(cihrfs.getRelativeChangeToGroups(changeObject)).toBe('edited');
```

Comment:
This is not really an edit right? Nothing changed in the object right?

Reply:
If `new_value` and `old_value` are not empty, then it was edited. I think this method is related to the value of `new_value` and `old_value` and not to its inside value(if it is an array or object).